### PR TITLE
feat(cli): headless record / export / info commands

### PR DESCRIPTION
**Pull request by PeterTakahashi**
## Summary

This PR adds a headless CLI to OpenScreen so scripts, CI pipelines, and AI coding agents can produce polished demos end-to-end — record, edit the project JSON programmatically, render — with no visible windows and machine-readable output:

```bash
# record the screen (native ScreenCaptureKit/WGC pipeline, cursor telemetry included)
openscreen record --duration 20 --project demo.openscreen --json

# agents can add zooms/annotations/trims by editing the .openscreen JSON directly

# render with the real export pipeline, optionally mixing in a TTS voiceover
openscreen export demo.openscreen -o demo.mp4 --audio voice.m4a --json
```

Full docs in `docs/cli.md` (commands, NDJSON protocol, exit codes, an end-to-end agent example).

## Why

`.openscreen` projects are plain JSON and the export pipeline is already proven headless by the `HEADLESS=true` e2e test — the missing piece was a way to drive both without the GUI. With this CLI, an AI coding agent can demo the product it just built: record it, zoom into the interesting parts, narrate with TTS, and ship an MP4.

## Design — deliberately minimal footprint

The CLI **reuses the existing pipelines unchanged** via the hidden-window pattern your `tests/e2e/gif-export.spec.ts` already established:

- `electron/cli/args.ts` — pure argv parser (14 unit tests). Skips leading Chromium switches so `Openscreen.AppImage --no-sandbox export …` works.
- `electron/cli/cliMain.ts` — headless boot: no HUD/tray/menu/dock, reuses `registerIpcHandlers` with inert window callbacks (the big handler module is untouched), NDJSON/`--json` protocol on stdout, diagnostics on stderr, SIGINT/SIGTERM/stdin `stop`, EPIPE-safe writes, exit codes 0/1/2, SwiftShader fallback for GPU-less environments.
- `src/cli/CliExportRunner.tsx` — loads the project via the existing `load-project-file-from-path` bridge action, rebuilds the same exporter config the editor's dialog builds, runs `VideoExporter`/`GifExporter`. DOM-derived preview dimensions are replaced by a deterministic 1280×720 reference box (`--preview-size` to override).
- `src/cli/CliRecordRunner.tsx` — drives `useScreenRecorder` (source by display index or window-title substring, mic by device label, system audio, cursor modes, `--duration`), then optionally writes a ready-to-export project file.
- `--audio` voiceover mixing (`src/lib/exporter/voiceoverMix.ts`): copies the exported video packets without re-encoding (mediabunny `Input`/`EncodedPacketSink` → `EncodedVideoPacketSource`) and renders the audio offline with `OfflineAudioContext` (`mix` or `replace`, `--audio-offset`).

Only three GUI-side touches: the `main.ts` CLI branch, new preload IPC methods, and exposing `startRecordingImmediately()` (countdown-less start) from `useScreenRecorder`.

## Tested (macOS 26.6, Apple Silicon)

- `record --duration 4/5` → MP4 + `.cursor.json` + session manifest; stop via SIGINT and via `--duration`; `--project` output loads in the GUI editor
- `export` MP4 (zooms, annotations, padding, multi-track audio fixture) and GIF (magic bytes + dimensions verified); output verified with ffprobe
- Full agent pipeline: record → programmatic JSON edit → export; and `--audio` with a real `say`-generated TTS track (volumedetect confirms audible narration)
- `openscreen export | head` (EPIPE) no longer crashes or shows Electron's error dialog; renderer console errors are forwarded to stderr
- `npm run lint` clean, `tsc` clean, all 441 vitest unit tests pass (14 new), GUI mode boots unaffected

## Platform caveats (honest disclosure)

Windows and Linux paths were **statically reviewed, not device-tested** (I only have macOS hardware). The review-driven fixes are included: leading-switch subcommand detection, stdin guard moved inside try (Windows GUI-subsystem stdio edge case), and platform notes in `docs/cli.md` (Windows: no SIGTERM — use stdin `stop`/`--duration`; Linux Wayland: PipeWire portal may prompt and headless sessions can't record). Happy to iterate if maintainers can test on those platforms.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a headless command-line interface for recording screens, exporting MP4/GIF files, listing capture sources, generating captions, inspecting projects, and packing portable projects.
  * Added machine-readable NDJSON output, progress reporting, validation, exit codes, and support for stopping recordings from the terminal.
  * Added optional voiceover mixing or replacement during MP4 export.
* **Documentation**
  * Added CLI setup, usage examples, workflows, options, and troubleshooting guidance.

### CODE REVIEW

**Comment by PeterTakahashi** [comment]
Addressed the CodeRabbit review in cc64b110 — all 8 actionable findings plus 6 of the nitpicks:

**Actionable**
- **voiceoverMix clipping**: `mix` mode now ducks the original bed to 0.4 gain by default (voiceover stays at unity), so the sum can't hard-clip; documented in `docs/cli.md`. Verified with volumedetect: max -2.5 dB on a mixed export.
- **voiceoverMix buffering**: the full-MP4 `arrayBuffer()` copy is now only taken in `mix` mode; `replace` skips it.
- **stop-before-start race**: an early SIGINT/stdin stop is remembered in `stopRequestedRef` and applied the moment recording flips on.
- **silent start failure**: since `useScreenRecorder` reports start failures via toast/console rather than rejecting, the record runner arms a 30s watchdog after `startRecordingImmediately()` and fails the run with a pointer to stderr (where renderer console errors are already forwarded). I kept the hook's semantics unchanged to avoid altering GUI behavior — happy to make it reject instead if you'd prefer that contract.
- **false failure on teardown**: `window-all-closed` now checks the `finished` flag before reporting.
- **info-command writes** now go through `safeWrite`; console rerouting stringification is guarded against circular args.
- **`--preview-size 0x0`** rejected.
- **docs fence language** added.

**Nitpicks taken**: metadata-probe 30s timeout, `Output.cancel()` on remux failure, typed preload CLI methods, win32-safe test path assertions, ref sync moved out of render.

Re-verified after the changes: full unit suite (439 tests), repo lint, `tsc`, and E2E smoke on macOS (record 3s + export with `--audio-mode mix`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by PeterTakahashi** [comment]
Pushed 6dcd7787 with four more commands that close the remaining gaps between the CLI and GUI workflows (docs updated in `docs/cli.md`):

- **`export --auto-zoom`** — runs the editor's cursor-dwell suggestion engine (`timeline/zoomSuggestionUtils`) over the recording's telemetry before rendering. CLI recordings previously never benefited from auto-zoom because suggestions are computed when the editor mounts; this brings the "it just looks good" behavior to headless exports. Verified on a real recording: 4 regions detected and rendered.
- **`sources`** — lists capturable displays, windows, and microphones (same enumeration as the GUI picker) so scripts can pick `--display`/`--window`/`--mic-device` values without trial and error.
- **`pack <project> --out <dir>`** — copies a project plus its media and cursor-telemetry sidecar into one portable folder. `getApprovedProjectSession` and the export runner gain a same-basename sibling fallback, so a packed folder keeps working after being moved (tested: packed, moved the folder, exported from the stale-path project successfully). The fallback only triggers when the stored path is missing and only considers files inside the already-trusted project directory, so the approval model is unchanged.
- **`captions <project>`** — on-device Whisper transcription written into the project as auto-caption annotations, mirroring `VideoEditor.generateAutoCaptions` (leading-silence trim, empty-retry, word grouping). Idempotent: re-running replaces earlier auto-captions and preserves manual annotations. Verified end-to-end: recorded → TTS voiceover mixed → captions generated → subtitles burned into the export.

All verified on macOS; 443 unit tests (7 new), lint and tsc clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019ceTJjoPJYa5BdmXZdDc5J

**Comment by PeterTakahashi** [comment]
Addressed the second review round — all three findings:

- `docs/cli.md`: the `sources --json` example now shows the real `done`-event payload as valid JSON, and the automated-demo example's wording now matches `--audio-mode replace` (with a note on the ducking default).
- `electron/cli/args.test.ts`: added the requested edge-case coverage for the `sources`, `pack`, and `captions` parsers — missing required paths, unknown options, extra positional arguments, and non-integer/zero word counts.

Lint and the full unit suite pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019ceTJjoPJYa5BdmXZdDc5J

**Comment by EtienneLescot** [comment]
**`main` has just been rebased onto the 1.8.0 line** — a ~390-commit merge landing the v4 editor, the native Rust compositor and export path, and the Linux capture stack. Every open PR is out of date; sorry for the churn. Here's where this one stands:

This one needs a bit more than a rebase. The model layer you build on survived — `@/components/video-editor/editorDefaults`, `projectPersistence` and `types` are all still there — but five imports are gone:

- `@/lib/exporter/videoExporter` and `@/lib/exporter/gifExporter` — the web MP4/GIF pipeline was replaced by the native compositor
- `@/components/video-editor/timeline/zoomSuggestionUtils`
- `@/lib/cursor/nativeCursor`
- `@/lib/captioning/leadingSilence`

`docs/cli.md`, `electron/cli/*` and the NDJSON contract are unaffected. Export now goes through `src/native/` into `crates/compositor`.

The design holds — it's the plumbing that moved. Happy to sketch the mapping if that helps.

**Comment by EtienneLescot** [comment]
Pushed the rebase to `contrib/pr-176-rebased` — your 4 commits replayed onto current `main`, all conflicts resolved (`preload.ts`, `electron-env.d.ts`, `App.tsx`, `package.json`, `.gitignore`).

Two things I dropped as leftovers from the old `main` rather than yours: the `test:browser` scripts (the vitest browser config is gone, and there are no `.browser.test.*` files left) and the `public/caption-assets` ignore (`fetch-caption-model.mjs` was deleted).

What's left is six symbols:

| import | where it went |
|---|---|
| `buildAutoZoomSuggestions` | `@/lib/ai-edition/timeline/zoom-suggestions` |
| `DEFAULT_ZOOM_DEPTH`, `ZOOM_DEPTH_SCALES` | `@/lib/ai-edition/timeline/zoom-scale` |
| `clampFocusToDepth` | deleted, no successor |
| `captionSegmentsToAnnotationRegions` | deleted, no successor |
| `trimLeadingSilenceMono16k` | deleted, no successor |
| `GifExporter` / `VideoExporter` | `exportGifNative` / `exportMultiNative` in `@/native` |

The last row is the real work: the native exporters take an `AxcutDocument` and write the file themselves instead of returning a Blob. `migrateProjectDataToAxcutDocument` (in `@/lib/ai-edition/document/migrate`) bridges from the project format you already load, so the `.openscreen` JSON contract — the point of the PR — doesn't have to change. But it does change what `--json` can report, and that's your call, so I stopped rather than guess at your CLI's contract.

**Comment by PeterTakahashi** [comment]
@EtienneLescot Thank you for the rebase branch and the symbol map — that table saved hours. The branch is now ported on top of `contrib/pr-176-rebased` (force-pushed here) and every command is verified end-to-end on macOS against the native pipeline.

**The six symbols:**
- `buildAutoZoomSuggestions` / zoom scale constants → repointed to `@/lib/ai-edition/timeline/zoom-suggestions` and `zoom-scale`; auto-zoom suggestions now append straight to `document.zoomRanges`.
- The three deleted helpers (`captionSegmentsToAnnotationRegions`, `trimLeadingSilenceMono16k`, `clampFocusToDepth` — plus `hasNativeCursorRecordingData`) are vendored under `src/cli/vendor/` with provenance headers, since the CLI still writes caption annotations into `.openscreen` projects (the v2 route your migration preserves).
- `VideoExporter`/`GifExporter` → the runner now migrates the project with `migrateProjectDataToAxcutDocument`, calls `applyProbedDuration` (without it the clip has no duration and the export is a single frame — worth a doc note somewhere), then mirrors the v4 ExportDialog exactly: same `buildNativeClipList` shape, same scene JSON, same crop-aware smallest-clip sizing, same GIF dimension capping.

**The `--json` contract call you left to me:** kept stable. The CLI still emits `progress {percentage, currentFrame, totalFrames, estimatedTimeRemaining}` — synthesized renderer-side from the `export:native-progress` frame counts with the total computed like the dialog does — and `done {outputPath, format, width, height}`. Two visible changes, both documented in `docs/cli.md`: `--preview-size` is now an accepted no-op (scene geometry is percentage-based), and `--audio` reads the natively-written file back, mixes the voiceover (mediabunny remux is untouched upstream, thankfully), and overwrites `outPath`.

Also registered `registerSttIpc` in CLI mode so headless `captions` works against the whisper.cpp engine — verified including the first-run model download.

Verified on macOS (Apple Silicon): export MP4/GIF through the Rust compositor, `--audio` mix, `--auto-zoom` (4 regions from real telemetry), record 3s, captions. `tsc` and lint clean (the 11 remaining warnings pre-exist on `main`); the `webm-seek-index` test failures also reproduce on a pristine `contrib/pr-176-rebased` checkout, so they're unrelated to this branch.

One setup note that may bite other contributors: on an Apple Silicon machine with a Rosetta-installed rustup (default host `x86_64-apple-darwin`), the compositor build fails in bindgen because Xcode's `libclang.dylib` is arm64-only. `rustup toolchain install stable-aarch64-apple-darwin` + `RUSTUP_TOOLCHAIN=stable-aarch64-apple-darwin npm run build:native:compositor:mac` fixes it without touching the default host.

#190 gets the same treatment next (multi-window mode re-added to the split `ScreenCaptureRecorder.swift`). I've also tried to enable "Allow edits by maintainers" there via the API without luck — will sort it out.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019ceTJjoPJYa5BdmXZdDc5J

**Comment by EtienneLescot** [comment]
﻿Pushed two commits to this branch (thanks for `maintainer_can_modify`):

**`17fde1a`** — three findings from the last review round:
- `parseExport` validated `--audio` before `--out` had resolved the format, so `--audio v.m4a -o out.gif` slipped through. Moved the check after inference + test.
- `pack` derived destinations from the basename alone: a webcam video sharing a name with the screen video overwrote it, and both project paths pointed at one file.
- `help`/`error` bypassed `safeWrite`, so `openscreen help | head` could throw before the EPIPE guards are installed.

**`c8b832b`** — `src/cli/vendor/captionRegions.ts` was `src/lib/captioning/annotationsFromCaptions.ts` copied whole (~460 identical lines) for the sake of the converter that 1.8 did delete. The copy had already lost `dedupeAdjacentCaptionRepeats` and `finalizeCaptionSegmentsForPlayback`, so CLI captions were drifting from the editor's. Only the converter is CLI-specific, so only it stays (`src/cli/captionAnnotations.ts`, 76 lines), importing the rest from the live module. Output is unchanged — the shared code was byte-identical, and the dropped `reconcileAutoCaptionTimelineGaps` was a no-op (its gap constant is 0). Small unit test added.

Verified on Windows: `tsc`, `tsc -p tsconfig.test.json`, `vite build`, biome and the unit tests are all clean, and `openscreen help` works.

Three things left before I'd merge:
1. ~3,300 lines of runtime with one test file (the argv parser). A single e2e in the existing `HEADLESS=true` harness — export a fixture project, assert the MP4 exists — would cover `cliMain` + `CliExportRunner` + the IPC contract at once.
2. `--preview-size` is documented as "accepted for compatibility" but is a no-op. Nothing has shipped yet, so there is no compatibility to keep — please drop the flag and its docs row.
3. The PR description is now stale: it describes `VideoExporter`/`GifExporter` and a deterministic 1280x720 preview box, but head goes through the native compositor.

Also noted, not blocking: `CliExportRunner` re-derives the export sizing that `ExportDialog` owns (the comments say as much) — worth extracting later so the two cannot drift. And `mixVoiceoverIntoVideo` holds the whole MP4 in memory; fine for demos, worth a caveat in the docs for long exports.

**Comment by EtienneLescot** [comment]
`885082eb` closes the two remaining points from my last comment.

**Tests.** `pack` and `info` only touch the project file and its media, but they sat inside `cliMain.ts` behind its `electron` import, so nothing could reach them — including the basename-collision fix I pushed in `17fde1af`. Moved to `electron/cli/projectCommands.ts` with the writer injected (nothing in there knows about `process.stdout` now), and covered: colliding screen/webcam basenames, the cursor sidecar, the stale-path sibling fallback, missing media, and `info`'s exit codes. Ablation-checked — reverting the de-dup turns the first test red, so it is testing what it claims to.

I did not add a Playwright e2e: `ci.yml` has no e2e job, so it would be a test nobody runs. `npm run test` (vitest) is what the Test job executes, so that is where the net belongs.

**`--preview-size` removed** — flag, contract fields, runner warning and both docs mentions.

Verified locally on Windows: `tsc`, `tsc -p tsconfig.test.json`, `vite build`, biome, `check-docs`, and 22 CLI unit tests green.

One thing left that is yours rather than mine: the PR description still describes `VideoExporter`/`GifExporter` and a deterministic 1280x720 preview box, which the native compositor replaced — worth a refresh before merge.

Note for whoever merges: CI has never actually run on this branch. The run on every head sits at `action_required` (fork PRs need a maintainer to approve workflows), so the only checks that have executed are CodeRabbit and the Discord sync. Please approve the workflow run and let it go green rather than trusting my local runs.

File: electron/cli/args.ts [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+			throw new Error(`--out must end in .mp4 or .gif, got "${request.outPath}"`);
+		}
+		const extFormat = ext === ".gif" ? "gif" : "mp4";
+		if (request.format && request.format !== extFormat) {
+			throw new Error(`--format ${request.format} conflicts with --out extension ${ext}`);
+		}
+		request.format = extFormat;
+	}
+	return request;
+}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/cli/args.ts:262]: Fixed in `17fde1af`: the `--audio`/gif check now runs after the `--out` extension has resolved `request.format`, so `--audio v.m4a -o out.gif` is rejected without an explicit `--format`. Covered by a new case in `args.test.ts`.

File: electron/cli/cliMain.ts [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+		return destination;
+	};
+
+	const screenSource = await resolveSource(screenVideoPath);
+	const newScreenPath = await copyIn(screenSource);
+
+	let newWebcamPath: string | undefined;
+	if (media.webcamVideoPath) {
+		newWebcamPath = await copyIn(await resolveSource(media.webcamVideoPath));
+	}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/cli/cliMain.ts]: Fixed in `17fde1af`: `copyIn` now de-duplicates the destination (`clip.mp4` / `clip-1.mp4`) instead of deriving it from the basename alone. `885082eb` adds the regression test — verified by ablation, reverting the de-dup turns it red. The command moved to `electron/cli/projectCommands.ts` so it could be tested at all.

File: electron/cli/cliMain.ts [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+	if (command.kind === "help") {
+		process.stdout.write(CLI_USAGE);
+		app.exit(0);
+		return;
+	}
+	if (command.kind === "error") {
+		process.stderr.write(`Error: ${command.message}\n\n${CLI_USAGE}`);
+		app.exit(2);
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/cli/cliMain.ts:226]: Fixed in `17fde1af`: both paths use `safeWrite` now.

File: electron/cli/cliMain.ts [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+
+	const packedProject: PackedProjectData = {
+		...data,
+		media: {
+			...media,
+			screenVideoPath: newScreenPath,
+			...(newWebcamPath ? { webcamVideoPath: newWebcamPath } : {}),
+		},
+	};
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/cli/cliMain.ts]: Not reproducible — skipping. `tsc --noEmit -p tsconfig.json` is clean on this branch, and `--listFiles` confirms `electron/cli/cliMain.ts` is in the program, so it is genuinely being checked. TypeScript's aliased-condition narrowing (4.4+) does narrow `media` through `const screenVideoPath = media?.screenVideoPath` because `media` is a `const`, which is why the later `media.webcamVideoPath` accesses type-check. The code has since moved to `electron/cli/projectCommands.ts` unchanged, still clean.

File: src/hooks/useScreenRecorder.ts [hunk 5]
```diff
@@ -54,6 +54,8 @@ type UseScreenRecorderReturn = {
 	saving: boolean;
 	elapsedSeconds: number;
 	toggleRecording: () => void;
+	/** Starts recording with no countdown overlay. Used by the headless CLI runner. */
+	startRecordingImmediately: () => Promise<void>;
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/hooks/useScreenRecorder.ts:59]: Addressed, but not the way suggested — leaving the reasoning here since this thread never got a reply.

Changing `startRecordingImmediately` to `Promise<boolean>` means changing the return contract of a hook the whole GUI shares, for a CLI-only need. The runner handles it locally instead, and covers two hang paths rather than one:

- `src/cli/CliRecordRunner.tsx` overrides `window.alert` before starting. `startRecording` reports some failures (no selected source) through a blocking `alert()`, which on a hidden window is an invisible modal that freezes the renderer forever — a failure mode the `Promise<boolean>` change would not have caught.
- A 30s watchdog fires `fail(...)` if `recording` never flips to `true`. The hook's `catch` calls `setRecording(false)`, so `recordingStartedAtRef` stays `null` and the deadline is reached; the CLI exits with an error instead of blocking.

The tradeoff is that a failed start surfaces after 30s with a generic message pointing at stderr, rather than failing fast. Acceptable for an error path, and it keeps the shared hook untouched.

**Review by EtienneLescot — APPROVED** [review]
